### PR TITLE
improve tests

### PR DESCRIPTION
**Pull request by jasmainak**
closes https://github.com/pavanramkumar/pyglmnet/issues/70

@pavanramkumar the `'poissonexp'` link function appears to be broken (?). Take a look at the `beta` in the test. They are way off from the simulated ones ...

### CODE REVIEW

**Comment by jasmainak** [comment]
betas

**Comment by pavanramkumar** [comment]
@jasmainak look at `notebooks/poissonexp_example.ipynb`. I'm comparing the test_`pyglmnet.py` code against this one line by line to look for discrepancies

**Comment by pavanramkumar** [comment]
Ok, firstly, to your `#FIX ME` comment, there is no special reason to use `reg_lambda = [0.05, 0.01]` other than the fact that I did not want to make the tests slow (running 10 vs. 2 times)

**Comment by pavanramkumar** [comment]
Secondly, it turns out that the core issue is related to warm restarts. Props to @hugoguh for testing this today!

```python
reg_lambda = np.logspace(np.log(0.5), np.log(0.01), 10, base=np.exp(1))
```
which gives
```python
array([ 0.5       ,  0.3237394 ,  0.2096144 ,  0.13572088,  0.08787639,
        0.0568981 ,  0.03684031,  0.02385332,  0.01544452,  0.01      ])
```
versus

```python
reg_lambda = [0.5, 0.01]
```

versus

```python
reg_lambda = [0.5, 0.01]
```
each, give vastly different results for `0.01` even though it is the same value of lambda.

I remember reading in the paper that even if you want to estimate the fit for one value of lambda, using a regularization path often improves things. In the paper they mention that is improves speed of convergence, but it appears that it also matters for accuracy of estimate. Warm restarts are most effective when previous estimate was done using a very similar but slightly larger value of lambda.

This probably gets into the nuts and bolts of the algorithm and brings up the issue of how to best choose the regularization path even if you know the optimal lambda.

**Comment by pavanramkumar** [comment]
I believe that the R package creates a regularization path under the hood even though the user may specify a single lambda. Need to verify this.

We can try to reproduce the problem by using `ElasticNetCV`: http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ElasticNetCV.html#sklearn.linear_model.ElasticNetCV

**Comment by pavanramkumar** [comment]
@jasmainak should i accept this PR for shape issues and move reg_lambda path to a separate thread? 

if so, would recommend modifying `test_pyglmnet` so that you don't need to set a special `reg_lambda` for `poissonexp`. only `learning_rate=1e-5` will do. i can also do this myself after merging. let me know.

**Comment by jasmainak** [comment]
@pavanramkumar ready to merge. 

[This](https://github.com/pavanramkumar/pyglmnet/pull/76#issuecomment-218638433) can be a very good test. In fact, I would even suggest that you write the test first (for what you expect according to your intuition) and then use `nosetests --pdb test_pyglmnet` to debug. It's faster that way than bug-hunting manually :)

File: .travis.yml [hunk 1]
```diff
@@ -14,6 +14,6 @@ install:
     - python setup.py install
 script:
     - nosetests . --with-coverage
-    - flake8 --count pyglmnet --exit-zero;
+    - flake8 --count pyglmnet
```
Comment (jasmainak) [inline .travis.yml]: @pavanramkumar I reverted this change because we really should conform to `flake8` conventions ...

Comment (pavanramkumar) [inline .travis.yml]: @jasmainak that's fine for now. just wanted to make sure developers don't feel discouraged to contribute because of flake8 norms in the future :)

Comment (jasmainak) [inline .travis.yml]: yes I figured ;) I guess it's a trade-off. If your code is easy to read (due to the conventions) and understand, then you will attract good developers. But then they will discover that the reason it's clean is partly due to the conventions :)
